packages/dapp/components/redeem/lib/use-prices.ts

```diff
@@ -1,28 +1,32 @@
-import useDeployedContracts from "@/components/lib/hooks/contracts/use-deployed-contracts";
-import useManagerManaged from "@/components/lib/hooks/contracts/use-manager-managed";
+import { getIMetaPoolContract } from "@/components/utils/contracts";
+import useProtocolContracts from "@/components/lib/hooks/contracts/use-protocol-contracts";
+import useWeb3 from "@/components/lib/hooks/use-web-3";
 import { BigNumber, utils } from "ethers";
 import { useEffect, useState } from "react";
 
 const usePrices = (): [BigNumber | null, BigNumber | null, () => Promise<void>] => {
-  const deployedContracts = useDeployedContracts();
-  const managedContracts = useManagerManaged();
+  const protocolContracts = useProtocolContracts();
+  const { provider } = useWeb3();
 
   const [twapPrice, setTwapPrice] = useState<BigNumber | null>(null);
   const [spotPrice, setSpotPrice] = useState<BigNumber | null>(null);
 
   async function refreshPrices() {
```

Comment:
This code leaves desire here, can yo do something like, improvde & debug?

```typescript
async function refreshPrices() {
  try {
    if (!managedContracts || !deployedContracts) {
      return;
    }

    const dollarTokenAddress = await deployedContracts.manager.dollarTokenAddress();
    const newTwapPrice = await managedContracts.dollarTwapOracle.consult(dollarTokenAddress);
    const newSpotPrice = await managedContracts.dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));

    if (protocolContracts && provider) {
      const dollar3poolMarket = await protocolContracts.managerFacet.stableSwapMetaPoolAddress();
      const dollarMetapool = getIMetaPoolContract(dollar3poolMarket, provider);
      const newTwapPrice3pool = await protocolContracts.twapOracleDollar3poolFacet.consult(dollarTokenAddress);
      const newSpotPrice3pool = await dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));

      setTwapPrice(newTwapPrice3pool);
      setSpotPrice(newSpotPrice3pool);
    }
  } catch (error) {
    console.error("Error in refreshPrices:", error);
  }
}
```

Reply:
> This code leaves desiring here, can yo do something like, improve & debug?
> 
> ```ts
> async function refreshPrices() {
>   try {
>     if (!managedContracts || !deployedContracts) {
>       return;
>     }
> 
>     const dollarTokenAddress = await deployedContracts.manager.dollarTokenAddress();
>     const newTwapPrice = await managedContracts.dollarTwapOracle.consult(dollarTokenAddress);
>     const newSpotPrice = await managedContracts.dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
> 
>     if (protocolContracts && provider) {
>       const dollar3poolMarket = await protocolContracts.managerFacet.stableSwapMetaPoolAddress();
>       const dollarMetapool = getIMetaPoolContract(dollar3poolMarket, provider);
>       const newTwapPrice3pool = await protocolContracts.twapOracleDollar3poolFacet.consult(dollarTokenAddress);
>       const newSpotPrice3pool = await dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
> 
>       setTwapPrice(newTwapPrice3pool);
>       setSpotPrice(newSpotPrice3pool);
>     }
>   } catch (error) {
>     console.error("Error in refreshPrices:", error);
>   }
> }
> ```

Improved it as recommended

Reply:
> This code leaves desiring here, can yo do something like, improve & debug?
> 
> ```ts
> async function refreshPrices() {
>   try {
>     if (!managedContracts || !deployedContracts) {
>       return;
>     }
> 
>     const dollarTokenAddress = await deployedContracts.manager.dollarTokenAddress();
>     const newTwapPrice = await managedContracts.dollarTwapOracle.consult(dollarTokenAddress);
>     const newSpotPrice = await managedContracts.dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
> 
>     if (protocolContracts && provider) {
>       const dollar3poolMarket = await protocolContracts.managerFacet.stableSwapMetaPoolAddress();
>       const dollarMetapool = getIMetaPoolContract(dollar3poolMarket, provider);
>       const newTwapPrice3pool = await protocolContracts.twapOracleDollar3poolFacet.consult(dollarTokenAddress);
>       const newSpotPrice3pool = await dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
> 
>       setTwapPrice(newTwapPrice3pool);
>       setSpotPrice(newSpotPrice3pool);
>     }
>   } catch (error) {
>     console.error("Error in refreshPrices:", error);
>   }
> }
> ```

I'm a bit confused.. I understand that `deployedContracts` and `managedContracts` should be replaced with `protcolContracts`.

Reply:
this is a bad approach:

protocolContracts.managerFacet && await protocolContracts.managerFacet.dollarTokenAddress();

please refactor according to the suggested code, if possible improve, and obviously debug it

Reply:
> this is a bad approach:
> 
> `protocolContracts.managerFacet && await protocolContracts.managerFacet.dollarTokenAddress();`
> 
> please refactor according to the suggested code, if possible improve, and obviously debug it

```
  async function refreshPrices() {
    try {
      if(!protocolContracts || !provider) {
        return;
      }

      if(protocolContracts.managerFacet && protocolContracts.twapOracleDollar3poolFacet) {
        const dollarTokenAddress = await protocolContracts.managerFacet.dollarTokenAddress();
        const newTwapPrice = await protocolContracts.twapOracleDollar3poolFacet.consult(dollarTokenAddress);
        const dollar3poolMarket = await protocolContracts.managerFacet.stableSwapMetaPoolAddress();
        const dollarMetapool = getIMetaPoolContract(dollar3poolMarket, provider)
        const newSpotPrice = await dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
        setTwapPrice(newTwapPrice);
        setSpotPrice(newSpotPrice);
      }

    } catch (error) {
      console.log("Error in refreshPrices: ", error)
    }
  }
```
So you want this approach instead of it?

Reply:
> > this is a bad approach:
> > `protocolContracts.managerFacet && await protocolContracts.managerFacet.dollarTokenAddress();`
> > please refactor according to the suggested code, if possible improve, and obviously debug it
> 
> ```
>   async function refreshPrices() {
>     try {
>       if(!protocolContracts || !provider) {
>         return;
>       }
> 
>       if(protocolContracts.managerFacet && protocolContracts.twapOracleDollar3poolFacet) {
>         const dollarTokenAddress = await protocolContracts.managerFacet.dollarTokenAddress();
>         const newTwapPrice = await protocolContracts.twapOracleDollar3poolFacet.consult(dollarTokenAddress);
>         const dollar3poolMarket = await protocolContracts.managerFacet.stableSwapMetaPoolAddress();
>         const dollarMetapool = getIMetaPoolContract(dollar3poolMarket, provider)
>         const newSpotPrice = await dollarMetapool["get_dy(int128,int128,uint256)"](0, 1, utils.parseEther("1"));
>         setTwapPrice(newTwapPrice);
>         setSpotPrice(newSpotPrice);
>       }
> 
>     } catch (error) {
>       console.log("Error in refreshPrices: ", error)
>     }
>   }
> ```
> 
> So you want this approach instead of it?

@molecula451 I'm waiting. @rndquu I need your opinion here too..

Reply:
take a look to this line https://github.com/ubiquity/ubiquity-dollar/blob/dd22fa0d1887b89953293a705a544250fcb6583a/packages/dapp/components/redeem/lib/use-prices.ts#L20

and take a look to the suggestions im providing, you'll be enlighten therefore the code shall be much more cleaner and readbale, reusable